diff -urN nginx-1.31.4/auto/cc/clang nginx-1.31.4/auto/cc/clang
--- nginx-1.31.4/auto/cc/clang	2026-08-20 18:25:06.242214766 +0000
+++ nginx-1.31.4/auto/cc/clang	2026-08-20 18:27:15.285187448 +0000
@@ -80,7 +80,7 @@
 fi

 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"

 # debug
 CFLAGS="$CFLAGS -g"
diff -urN nginx-1.31.4/auto/cc/conf nginx-1.31.4/auto/cc/conf
--- nginx-1.31.4/auto/cc/conf	2026-08-20 18:25:06.246066256 +0000
+++ nginx-1.31.4/auto/cc/conf	2026-08-20 18:27:15.409065091 +0000
@@ -144,7 +144,7 @@
 CFLAGS="$CFLAGS $NGX_CC_OPT"
 NGX_TEST_LD_OPT="$NGX_LD_OPT"

-if [ "$NGX_PLATFORM" != win32 ]; then
+if [ 1 ]; then

     if test -n "$NGX_LD_OPT"; then
         ngx_feature=--with-ld-opt=\"$NGX_LD_OPT\"
diff -urN nginx-1.31.4/auto/cc/gcc nginx-1.31.4/auto/cc/gcc
--- nginx-1.31.4/auto/cc/gcc	2026-08-20 18:25:06.243183851 +0000
+++ nginx-1.31.4/auto/cc/gcc	2026-08-20 18:27:15.285934630 +0000
@@ -158,7 +158,7 @@


 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"

 # debug
 CFLAGS="$CFLAGS -g"
diff -urN nginx-1.31.4/auto/cc/icc nginx-1.31.4/auto/cc/icc
--- nginx-1.31.4/auto/cc/icc	2026-08-20 18:25:06.243646028 +0000
+++ nginx-1.31.4/auto/cc/icc	2026-08-20 18:27:15.286299765 +0000
@@ -103,7 +103,7 @@
 esac

 # stop on warning
-CFLAGS="$CFLAGS -Werror"
+#CFLAGS="$CFLAGS -Werror"

 # debug
 CFLAGS="$CFLAGS -g"
diff -urN nginx-1.31.4/auto/options nginx-1.31.4/auto/options
--- nginx-1.31.4/auto/options	2026-08-20 18:25:06.230358292 +0000
+++ nginx-1.31.4/auto/options	2026-08-20 18:27:15.395065833 +0000
@@ -384,6 +384,7 @@
         --with-pcre)                     USE_PCRE=YES               ;;
         --with-pcre=*)                   PCRE="$value"              ;;
         --with-pcre-opt=*)               PCRE_OPT="$value"          ;;
+        --with-pcre-conf-opt=*)          PCRE_CONF_OPT="$value"     ;;
         --with-pcre-jit)                 PCRE_JIT=YES               ;;
         --without-pcre2)                 PCRE2=DISABLED             ;;

@@ -612,6 +613,7 @@
   --with-pcre                        force PCRE library usage
   --with-pcre=DIR                    set path to PCRE library sources
   --with-pcre-opt=OPTIONS            set additional build options for PCRE
+  --with-pcre-conf-opt=OPTIONS       set additional configure options for PCRE
   --with-pcre-jit                    build PCRE with JIT compilation support
   --without-pcre2                    do not use PCRE2 library

diff -urN nginx-1.31.4/src/core/nginx.c nginx-1.31.4/src/core/nginx.c
--- nginx-1.31.4/src/core/nginx.c	2026-08-20 18:25:06.281397255 +0000
+++ nginx-1.31.4/src/core/nginx.c	2026-08-20 18:26:34.814546860 +0000
@@ -188,6 +188,7 @@
 static u_char      *ngx_conf_file;
 static u_char      *ngx_conf_params;
 static char        *ngx_signal;
+ngx_pool_t         *saved_init_cycle_pool = NULL;


 static char **ngx_os_environ;
@@ -256,6 +257,8 @@
         return 1;
     }

+    saved_init_cycle_pool = init_cycle.pool;
+
     if (ngx_save_argv(&init_cycle, argc, argv) != NGX_OK) {
         return 1;
     }
@@ -1113,6 +1116,8 @@

     ccf->daemon = NGX_CONF_UNSET;
     ccf->master = NGX_CONF_UNSET;
+    ccf->privileged_agent = NGX_CONF_UNSET;
+    ccf->privileged_agent_connections = NGX_CONF_UNSET_UINT;
     ccf->timer_resolution = NGX_CONF_UNSET_MSEC;
     ccf->shutdown_timeout = NGX_CONF_UNSET_MSEC;

@@ -1142,6 +1147,8 @@

     ngx_conf_init_value(ccf->daemon, 1);
     ngx_conf_init_value(ccf->master, 1);
+    ngx_conf_init_value(ccf->privileged_agent, 0);
+    ngx_conf_init_uint_value(ccf->privileged_agent_connections, 512);
     ngx_conf_init_msec_value(ccf->timer_resolution, 0);
     ngx_conf_init_msec_value(ccf->shutdown_timeout, 0);

diff -urN nginx-1.31.4/src/core/ngx_core.h nginx-1.31.4/src/core/ngx_core.h
--- nginx-1.31.4/src/core/ngx_core.h	2026-08-20 18:25:06.271793596 +0000
+++ nginx-1.31.4/src/core/ngx_core.h	2026-08-20 18:26:02.641199911 +0000
@@ -121,4 +121,6 @@
 #define NGX_DISABLE_SYMLINKS_NOTOWNER   2
 #endif

+extern ngx_pool_t        *saved_init_cycle_pool;
+
 #endif /* _NGX_CORE_H_INCLUDED_ */
diff -urN nginx-1.31.4/src/core/ngx_cycle.c nginx-1.31.4/src/core/ngx_cycle.c
--- nginx-1.31.4/src/core/ngx_cycle.c	2026-08-20 18:25:06.272666639 +0000
+++ nginx-1.31.4/src/core/ngx_cycle.c	2026-08-20 18:26:02.642222457 +0000
@@ -782,6 +782,10 @@

     if (ngx_process == NGX_PROCESS_MASTER || ngx_is_init_cycle(old_cycle)) {

+        if (ngx_is_init_cycle(old_cycle)) {
+            saved_init_cycle_pool = NULL;
+        }
+
         ngx_destroy_pool(old_cycle->pool);
         cycle->old_cycle = NULL;

diff -urN nginx-1.31.4/src/core/ngx_cycle.h nginx-1.31.4/src/core/ngx_cycle.h
--- nginx-1.31.4/src/core/ngx_cycle.h	2026-08-20 18:25:06.261610617 +0000
+++ nginx-1.31.4/src/core/ngx_cycle.h	2026-08-20 18:27:15.206293773 +0000
@@ -22,9 +22,17 @@
 #define NGX_DEBUG_POINTS_ABORT  2


+#define HAVE_PRIVILEGED_PROCESS_PATCH   1
+
+
+#define HAVE_INTERCEPT_ERROR_LOG_PATCH
+
+
 typedef struct ngx_shm_zone_s  ngx_shm_zone_t;

 typedef ngx_int_t (*ngx_shm_zone_init_pt) (ngx_shm_zone_t *zone, void *data);
+typedef ngx_int_t (*ngx_log_intercept_pt) (ngx_log_t *log, ngx_uint_t level,
+    u_char *buf, size_t len);

 struct ngx_shm_zone_s {
     void                     *data;
@@ -83,12 +91,18 @@
     ngx_str_t                 error_log;
     ngx_str_t                 lock_file;
     ngx_str_t                 hostname;
+
+    ngx_log_intercept_pt      intercept_error_log_handler;
+    void                     *intercept_error_log_data;
+    unsigned                  entered_logger;    /* :1 */
 };


 typedef struct {
     ngx_flag_t                daemon;
     ngx_flag_t                master;
+    ngx_flag_t                privileged_agent;
+    ngx_uint_t                privileged_agent_connections;

     ngx_msec_t                timer_resolution;
     ngx_msec_t                shutdown_timeout;
diff -urN nginx-1.31.4/src/core/ngx_log.c nginx-1.31.4/src/core/ngx_log.c
--- nginx-1.31.4/src/core/ngx_log.c	2026-08-20 18:25:06.260839205 +0000
+++ nginx-1.31.4/src/core/ngx_log.c	2026-08-20 18:27:15.207448881 +0000
@@ -112,6 +112,8 @@
     ngx_uint_t   wrote_stderr, debug_connection;
     u_char       errstr[NGX_MAX_ERROR_STR];

+    ngx_log_intercept_pt    log_intercept = NULL;
+
     last = errstr + NGX_MAX_ERROR_STR;

     p = ngx_cpymem(errstr, ngx_cached_err_log_time.data,
@@ -153,6 +155,16 @@
         p = last - NGX_LINEFEED_SIZE;
     }

+    if (ngx_cycle) {
+        log_intercept = ngx_cycle->intercept_error_log_handler;
+    }
+
+    if (log_intercept && !ngx_cycle->entered_logger) {
+        ngx_cycle->entered_logger = 1;
+        log_intercept(log, level, errstr, p - errstr);
+        ngx_cycle->entered_logger = 0;
+    }
+
     ngx_linefeed(p);

     wrote_stderr = 0;
diff -urN nginx-1.31.4/src/core/ngx_log.h nginx-1.31.4/src/core/ngx_log.h
--- nginx-1.31.4/src/core/ngx_log.h	2026-08-20 18:25:06.272393985 +0000
+++ nginx-1.31.4/src/core/ngx_log.h	2026-08-20 18:27:15.378388164 +0000
@@ -76,7 +76,9 @@
 };


-#define NGX_MAX_ERROR_STR   2048
+#ifndef NGX_MAX_ERROR_STR
+#define NGX_MAX_ERROR_STR   4096
+#endif


 /*********************************/
diff -urN nginx-1.31.4/src/core/ngx_resolver.c nginx-1.31.4/src/core/ngx_resolver.c
--- nginx-1.31.4/src/core/ngx_resolver.c	2026-08-20 18:25:06.268809929 +0000
+++ nginx-1.31.4/src/core/ngx_resolver.c	2026-08-20 18:26:02.596397105 +0000
@@ -9,12 +9,26 @@
 #include <ngx_core.h>
 #include <ngx_event.h>

+#if !(NGX_WIN32)
+#include <resolv.h>
+#endif
+

 #define NGX_RESOLVER_UDP_SIZE   4096

 #define NGX_RESOLVER_TCP_RSIZE  (2 + 65535)
 #define NGX_RESOLVER_TCP_WSIZE  8192

+#if !(NGX_WIN32)
+/*
+ * note that 2KB should be more than enough for majority of the
+ * resolv.conf files out there. it also acts as a safety guard to prevent
+ * abuse.
+ */
+#define NGX_RESOLVER_FILE_BUF_SIZE  2048
+#define NGX_RESOLVER_FILE_NAME      "/etc/resolv.conf"
+#endif
+

 typedef struct {
     u_char  ident_hi;
@@ -129,6 +143,191 @@
 #endif


+#if !(NGX_WIN32)
+static ngx_int_t
+ngx_resolver_read_resolv_conf(ngx_conf_t *cf, ngx_resolver_t *r, u_char *path,
+    size_t path_len)
+{
+    ngx_url_t                        u;
+    ngx_resolver_connection_t       *rec;
+    ngx_fd_t                         fd;
+    ngx_file_t                       file;
+    u_char                           buf[NGX_RESOLVER_FILE_BUF_SIZE];
+    u_char                           ipv6_buf[NGX_INET6_ADDRSTRLEN];
+    ngx_uint_t                       address = 0, j, total = 0;
+    ssize_t                          n, i;
+    enum {
+        sw_nameserver,
+        sw_spaces,
+        sw_address,
+        sw_skip
+    } state;
+
+    file.name.data = path;
+    file.name.len = path_len;
+
+    if (ngx_conf_full_name(cf->cycle, &file.name, 1) != NGX_OK) {
+        return NGX_ERROR;
+    }
+
+    fd = ngx_open_file(file.name.data, NGX_FILE_RDONLY,
+                       NGX_FILE_OPEN, 0);
+
+    if (fd == NGX_INVALID_FILE) {
+        ngx_conf_log_error(NGX_LOG_EMERG, cf, ngx_errno,
+                           ngx_open_file_n " \"%s\" failed", file.name.data);
+
+        return NGX_ERROR;
+    }
+
+    ngx_memzero(&file, sizeof(ngx_file_t));
+
+    file.fd = fd;
+    file.log = cf->log;
+
+    state = sw_nameserver;
+
+    n = ngx_read_file(&file, buf, NGX_RESOLVER_FILE_BUF_SIZE, 0);
+
+    if (n == NGX_ERROR) {
+        ngx_conf_log_error(NGX_LOG_ALERT, cf, ngx_errno,
+                           ngx_read_file_n " \"%s\" failed", file.name.data);
+    }
+
+    if (ngx_close_file(file.fd) == NGX_FILE_ERROR) {
+        ngx_conf_log_error(NGX_LOG_ALERT, cf, ngx_errno,
+                           ngx_close_file_n " \"%s\" failed", file.name.data);
+    }
+
+    if (n == NGX_ERROR) {
+        return NGX_ERROR;
+    }
+
+    if (n == 0) {
+        return NGX_OK;
+    }
+
+    for (i = 0; i < n && total < MAXNS; /* void */) {
+        if (buf[i] == '#' || buf[i] == ';') {
+            state = sw_skip;
+        }
+
+        switch (state) {
+
+        case sw_nameserver:
+
+            if ((size_t) n - i >= sizeof("nameserver") - 1
+                && ngx_memcmp(buf + i, "nameserver",
+                              sizeof("nameserver") - 1) == 0)
+            {
+                state = sw_spaces;
+                i += sizeof("nameserver") - 1;
+
+                continue;
+            }
+
+            break;
+
+        case sw_spaces:
+            if (buf[i] != '\t' && buf[i] != ' ') {
+                address = i;
+                state = sw_address;
+            }
+
+            break;
+
+        case sw_address:
+
+            if (buf[i] == CR || buf[i] == LF || i == n - 1) {
+                ngx_memzero(&u, sizeof(ngx_url_t));
+
+                u.url.data = buf + address;
+
+                if (i == n - 1 && buf[i] != CR && buf[i] != LF) {
+                    u.url.len = n - address;
+
+                } else {
+                    u.url.len = i - address;
+                }
+
+                u.default_port = 53;
+
+                /* IPv6? */
+                if (ngx_strlchr(u.url.data, u.url.data + u.url.len,
+                                ':') != NULL)
+                {
+                    if (u.url.len + 2 > sizeof(ipv6_buf)) {
+                        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                                           "IPv6 resolver address is too long:"
+                                           " \"%V\"", &u.url);
+
+                        return NGX_ERROR;
+                    }
+
+                    ipv6_buf[0] = '[';
+                    ngx_memcpy(ipv6_buf + 1, u.url.data, u.url.len);
+                    ipv6_buf[u.url.len + 1] = ']';
+
+                    u.url.data = ipv6_buf;
+                    u.url.len = u.url.len + 2;
+                }
+
+                if (ngx_parse_url(cf->pool, &u) != NGX_OK) {
+                    if (u.err) {
+                        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                                           "%s in resolver \"%V\"",
+                                           u.err, &u.url);
+                    }
+
+                    return NGX_ERROR;
+                }
+
+                rec = ngx_array_push_n(&r->connections, u.naddrs);
+                if (rec == NULL) {
+                    return NGX_ERROR;
+                }
+
+                ngx_memzero(rec, u.naddrs * sizeof(ngx_resolver_connection_t));
+
+                for (j = 0; j < u.naddrs; j++) {
+                    rec[j].sockaddr = u.addrs[j].sockaddr;
+                    rec[j].socklen = u.addrs[j].socklen;
+                    rec[j].server = u.addrs[j].name;
+                    rec[j].resolver = r;
+                }
+
+                total++;
+
+#if (NGX_DEBUG)
+                /*
+                 * logs with level below NGX_LOG_NOTICE will not be printed
+                 * in this early phase
+                 */
+                ngx_conf_log_error(NGX_LOG_NOTICE, cf, 0,
+                                   "parsed a resolver: \"%V\"", &u.url);
+#endif
+
+                state = sw_nameserver;
+            }
+
+            break;
+
+        case sw_skip:
+            if (buf[i] == CR || buf[i] == LF) {
+                state = sw_nameserver;
+            }
+
+            break;
+        }
+
+        i++;
+    }
+
+    return NGX_OK;
+}
+#endif
+
+
 ngx_resolver_t *
 ngx_resolver_create(ngx_conf_t *cf, ngx_str_t *names, ngx_uint_t n)
 {
@@ -259,6 +458,39 @@
             }

             continue;
+        }
+#endif
+
+#if !(NGX_WIN32)
+        if (ngx_strncmp(names[i].data, "local=", 6) == 0) {
+
+            if (ngx_strcmp(&names[i].data[6], "on") == 0) {
+                if (ngx_resolver_read_resolv_conf(cf, r,
+                                                  (u_char *)
+                                                  NGX_RESOLVER_FILE_NAME,
+                                                  sizeof(NGX_RESOLVER_FILE_NAME)
+                                                  - 1)
+                    != NGX_OK)
+                {
+                    ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                                       "unable to parse local resolver");
+                    return NULL;
+                }
+
+            } else if (ngx_strcmp(&names[i].data[6], "off") != 0) {
+                if (ngx_resolver_read_resolv_conf(cf, r,
+                                                  &names[i].data[6],
+                                                  names[i].len - 6)
+                    != NGX_OK)
+                {
+                    ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
+                                       "unable to parse local resolver");
+                    return NULL;
+                }
+
+            }
+
+            continue;
         }
 #endif

diff -urN nginx-1.31.4/src/core/ngx_thread_pool.c nginx-1.31.4/src/core/ngx_thread_pool.c
--- nginx-1.31.4/src/core/ngx_thread_pool.c	2026-08-20 18:25:06.260158285 +0000
+++ nginx-1.31.4/src/core/ngx_thread_pool.c	2026-08-20 18:27:15.154245516 +0000
@@ -587,7 +587,8 @@
     ngx_thread_pool_conf_t   *tcf;

     if (ngx_process != NGX_PROCESS_WORKER
-        && ngx_process != NGX_PROCESS_SINGLE)
+        && ngx_process != NGX_PROCESS_SINGLE
+        && !ngx_is_privileged_agent)
     {
         return NGX_OK;
     }
diff -urN nginx-1.31.4/src/event/ngx_event.c nginx-1.31.4/src/event/ngx_event.c
--- nginx-1.31.4/src/event/ngx_event.c	2026-08-20 18:25:06.345420930 +0000
+++ nginx-1.31.4/src/event/ngx_event.c	2026-08-20 18:26:02.693283070 +0000
@@ -197,6 +197,9 @@
     ngx_uint_t  flags;
     ngx_msec_t  timer, delta;

+    ngx_queue_t     *q;
+    ngx_event_t     *ev;
+
     if (ngx_timer_resolution) {
         timer = NGX_TIMER_INFINITE;
         flags = 0;
@@ -216,6 +219,13 @@
 #endif
     }

+    if (!ngx_queue_empty(&ngx_posted_delayed_events)) {
+        ngx_log_debug0(NGX_LOG_DEBUG_EVENT, cycle->log, 0,
+                       "posted delayed event queue not empty"
+                       " making poll timeout 0");
+        timer = 0;
+    }
+
     if (ngx_use_accept_mutex) {
         if (ngx_accept_disabled > 0) {
             ngx_accept_disabled--;
@@ -261,6 +271,35 @@
     ngx_event_expire_timers();

     ngx_event_process_posted(cycle, &ngx_posted_events);
+
+    while (!ngx_queue_empty(&ngx_posted_delayed_events)) {
+        q = ngx_queue_head(&ngx_posted_delayed_events);
+
+        ev = ngx_queue_data(q, ngx_event_t, queue);
+        if (ev->delayed) {
+            /* start of newly inserted nodes */
+            for (/* void */;
+                 q != ngx_queue_sentinel(&ngx_posted_delayed_events);
+                 q = ngx_queue_next(q))
+            {
+                ev = ngx_queue_data(q, ngx_event_t, queue);
+                ev->delayed = 0;
+
+                ngx_log_debug1(NGX_LOG_DEBUG_EVENT, cycle->log, 0,
+                               "skipping delayed posted event %p,"
+                               " till next iteration", ev);
+            }
+
+            break;
+        }
+
+        ngx_log_debug1(NGX_LOG_DEBUG_EVENT, cycle->log, 0,
+                       "delayed posted event %p", ev);
+
+        ngx_delete_posted_event(ev);
+
+        ev->handler(ev);
+    }
 }


@@ -671,6 +710,7 @@
     ngx_queue_init(&ngx_posted_accept_events);
     ngx_queue_init(&ngx_posted_next_events);
     ngx_queue_init(&ngx_posted_events);
+    ngx_queue_init(&ngx_posted_delayed_events);

     if (ngx_event_timer_init(cycle->log) == NGX_ERROR) {
         return NGX_ERROR;
diff -urN nginx-1.31.4/src/event/ngx_event_openssl.c nginx-1.31.4/src/event/ngx_event_openssl.c
--- nginx-1.31.4/src/event/ngx_event_openssl.c	2026-08-20 18:25:06.344545326 +0000
+++ nginx-1.31.4/src/event/ngx_event_openssl.c	2026-08-20 18:28:43.040812191 +0000
@@ -2317,6 +2317,31 @@
         return NGX_AGAIN;
     }

+#if OPENSSL_VERSION_NUMBER >= 0x10002000L
+    if (sslerr == SSL_ERROR_WANT_X509_LOOKUP
+#if defined(SSL_ERROR_PENDING_SESSION)
+        || sslerr == SSL_ERROR_PENDING_SESSION
+#endif
+#if defined(SSL_ERROR_WANT_CLIENT_HELLO_CB)
+        || sslerr == SSL_ERROR_WANT_CLIENT_HELLO_CB
+#endif
+       )
+    {
+        c->read->handler = ngx_ssl_handshake_handler;
+        c->write->handler = ngx_ssl_handshake_handler;
+
+        if (ngx_handle_read_event(c->read, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        if (ngx_handle_write_event(c->write, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        return NGX_AGAIN;
+    }
+#endif
+
     err = (sslerr == SSL_ERROR_SYSCALL) ? ngx_errno : 0;

     c->ssl->no_wait_shutdown = 1;
@@ -2464,6 +2489,55 @@
         return NGX_AGAIN;
     }

+    if (sslerr == SSL_ERROR_WANT_X509_LOOKUP) {
+        c->read->handler = ngx_ssl_handshake_handler;
+        c->write->handler = ngx_ssl_handshake_handler;
+
+        if (ngx_handle_read_event(c->read, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        if (ngx_handle_write_event(c->write, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        return NGX_AGAIN;
+    }
+
+#ifdef SSL_ERROR_PENDING_SESSION
+    if (sslerr == SSL_ERROR_PENDING_SESSION) {
+        c->read->handler = ngx_ssl_handshake_handler;
+        c->write->handler = ngx_ssl_handshake_handler;
+
+        if (ngx_handle_read_event(c->read, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        if (ngx_handle_write_event(c->write, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        return NGX_AGAIN;
+    }
+#endif
+
+#ifdef SSL_ERROR_WANT_CLIENT_HELLO_CB
+    if (sslerr == SSL_ERROR_WANT_CLIENT_HELLO_CB) {
+        c->read->handler = ngx_ssl_handshake_handler;
+        c->write->handler = ngx_ssl_handshake_handler;
+
+        if (ngx_handle_read_event(c->read, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        if (ngx_handle_write_event(c->write, 0) != NGX_OK) {
+            return NGX_ERROR;
+        }
+
+        return NGX_AGAIN;
+    }
+#endif
+
     err = (sslerr == SSL_ERROR_SYSCALL) ? ngx_errno : 0;

     c->ssl->no_wait_shutdown = 1;
diff -urN nginx-1.31.4/src/event/ngx_event_posted.c nginx-1.31.4/src/event/ngx_event_posted.c
--- nginx-1.31.4/src/event/ngx_event_posted.c	2026-08-20 18:25:06.345246987 +0000
+++ nginx-1.31.4/src/event/ngx_event_posted.c	2026-08-20 18:26:02.693789007 +0000
@@ -13,6 +13,7 @@
 ngx_queue_t  ngx_posted_accept_events;
 ngx_queue_t  ngx_posted_next_events;
 ngx_queue_t  ngx_posted_events;
+ngx_queue_t  ngx_posted_delayed_events;


 void
diff -urN nginx-1.31.4/src/event/ngx_event_posted.h nginx-1.31.4/src/event/ngx_event_posted.h
--- nginx-1.31.4/src/event/ngx_event_posted.h	2026-08-20 18:25:06.342820109 +0000
+++ nginx-1.31.4/src/event/ngx_event_posted.h	2026-08-20 18:26:02.694161373 +0000
@@ -45,6 +45,9 @@
 extern ngx_queue_t  ngx_posted_accept_events;
 extern ngx_queue_t  ngx_posted_next_events;
 extern ngx_queue_t  ngx_posted_events;
+extern ngx_queue_t  ngx_posted_delayed_events;
+
+#define HAVE_POSTED_DELAYED_EVENTS_PATCH


 #endif /* _NGX_EVENT_POSTED_H_INCLUDED_ */
diff -urN nginx-1.31.4/src/http/modules/ngx_http_log_module.c nginx-1.31.4/src/http/modules/ngx_http_log_module.c
--- nginx-1.31.4/src/http/modules/ngx_http_log_module.c	2026-08-20 18:25:06.307399909 +0000
+++ nginx-1.31.4/src/http/modules/ngx_http_log_module.c	2026-08-20 18:27:15.309248241 +0000
@@ -79,6 +79,8 @@
     time_t                      open_file_cache_valid;
     ngx_uint_t                  open_file_cache_min_uses;

+    ngx_flag_t                  escape_non_ascii;
+
     ngx_uint_t                  off;        /* unsigned  off:1 */
 } ngx_http_log_loc_conf_t;

@@ -136,7 +138,8 @@
     uintptr_t data);
 static u_char *ngx_http_log_variable(ngx_http_request_t *r, u_char *buf,
     u_char *end, ngx_http_log_op_t *op);
-static uintptr_t ngx_http_log_escape(u_char *dst, u_char *src, size_t size);
+static uintptr_t ngx_http_log_escape(ngx_http_log_loc_conf_t *lcf, u_char *dst,
+    u_char *src, size_t size);
 static size_t ngx_http_log_json_variable_getlen(ngx_http_request_t *r,
     uintptr_t data);
 static u_char *ngx_http_log_json_variable(ngx_http_request_t *r, u_char *buf,
@@ -189,6 +192,13 @@
       0,
       NULL },

+
+    { ngx_string("log_escape_non_ascii"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_CONF_FLAG,
+      ngx_conf_set_flag_slot,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      offsetof(ngx_http_log_loc_conf_t, escape_non_ascii),
+      NULL },
       ngx_null_command
 };

@@ -1017,6 +1027,7 @@
 ngx_http_log_variable_getlen(ngx_http_request_t *r, uintptr_t data)
 {
     uintptr_t                   len;
+    ngx_http_log_loc_conf_t    *lcf;
     ngx_http_variable_value_t  *value;

     value = ngx_http_get_indexed_variable(r, data);
@@ -1025,7 +1036,9 @@
         return 1;
     }

-    len = ngx_http_log_escape(NULL, value->data, value->len);
+    lcf = ngx_http_get_module_loc_conf(r, ngx_http_log_module);
+
+    len = ngx_http_log_escape(lcf, NULL, value->data, value->len);

     value->escape = len ? 1 : 0;

@@ -1037,6 +1050,7 @@
 ngx_http_log_variable(ngx_http_request_t *r, u_char *buf, u_char *end,
     ngx_http_log_op_t *op)
 {
+    ngx_http_log_loc_conf_t    *lcf;
     uintptr_t                   len;
     ngx_http_variable_value_t  *value;

@@ -1059,7 +1073,8 @@
         return ngx_cpymem(buf, value->data, value->len);

     } else {
-        len = ngx_http_log_escape(NULL, value->data, value->len);
+        lcf = ngx_http_get_module_loc_conf(r, ngx_http_log_module);
+        len = ngx_http_log_escape(lcf, NULL, value->data, value->len);

         if (ngx_http_log_check_length(r, buf, end, value->len + len * 3)
             != NGX_OK)
@@ -1067,16 +1082,17 @@
             return NULL;
         }

-        return (u_char *) ngx_http_log_escape(buf, value->data, value->len);
+        return (u_char *) ngx_http_log_escape(lcf, buf, value->data, value->len);
     }
 }


 static uintptr_t
-ngx_http_log_escape(u_char *dst, u_char *src, size_t size)
+ngx_http_log_escape(ngx_http_log_loc_conf_t *lcf, u_char *dst, u_char *src,
+    size_t size)
 {
-    ngx_uint_t      n;
-    static u_char   hex[] = "0123456789ABCDEF";
+    ngx_uint_t                   n;
+    static u_char                hex[] = "0123456789ABCDEF";

     static uint32_t   escape[] = {
         0xffffffff, /* 1111 1111 1111 1111  1111 1111 1111 1111 */
@@ -1096,6 +1112,13 @@
         0xffffffff, /* 1111 1111 1111 1111  1111 1111 1111 1111 */
     };

+    if (lcf->escape_non_ascii) {
+        ngx_memset(&escape[4], 0xff, sizeof(uint32_t) * 4);
+
+    } else {
+        ngx_memzero(&escape[4], sizeof(uint32_t) * 4);
+    }
+

     if (dst == NULL) {

@@ -1284,6 +1307,7 @@
     }

     conf->open_file_cache = NGX_CONF_UNSET_PTR;
+    conf->escape_non_ascii = NGX_CONF_UNSET;

     return conf;
 }
@@ -1299,6 +1323,8 @@
     ngx_http_log_fmt_t        *fmt;
     ngx_http_log_main_conf_t  *lmcf;

+    ngx_conf_merge_value(conf->escape_non_ascii, prev->escape_non_ascii, 1);
+
     if (conf->open_file_cache == NGX_CONF_UNSET_PTR) {

         conf->open_file_cache = prev->open_file_cache;
diff -urN nginx-1.31.4/src/http/modules/ngx_http_proxy_module.c nginx-1.31.4/src/http/modules/ngx_http_proxy_module.c
--- nginx-1.31.4/src/http/modules/ngx_http_proxy_module.c	2026-08-20 18:25:06.322883399 +0000
+++ nginx-1.31.4/src/http/modules/ngx_http_proxy_module.c	2026-08-20 18:27:15.337685723 +0000
@@ -813,13 +813,13 @@
 static ngx_http_variable_t  ngx_http_proxy_vars[] = {

     { ngx_string("proxy_host"), NULL, ngx_http_proxy_host_variable, 0,
-      NGX_HTTP_VAR_CHANGEABLE|NGX_HTTP_VAR_NOCACHEABLE|NGX_HTTP_VAR_NOHASH, 0 },
+      NGX_HTTP_VAR_CHANGEABLE|NGX_HTTP_VAR_NOCACHEABLE, 0 },

     { ngx_string("proxy_port"), NULL, ngx_http_proxy_port_variable, 0,
-      NGX_HTTP_VAR_CHANGEABLE|NGX_HTTP_VAR_NOCACHEABLE|NGX_HTTP_VAR_NOHASH, 0 },
+      NGX_HTTP_VAR_CHANGEABLE|NGX_HTTP_VAR_NOCACHEABLE, 0 },

     { ngx_string("proxy_add_x_forwarded_for"), NULL,
-      ngx_http_proxy_add_x_forwarded_for_variable, 0, NGX_HTTP_VAR_NOHASH, 0 },
+      ngx_http_proxy_add_x_forwarded_for_variable, 0, 0, 0 },

 #if 0
     { ngx_string("proxy_add_via"), NULL, NULL, 0, NGX_HTTP_VAR_NOHASH, 0 },
diff -urN nginx-1.31.4/src/http/ngx_http_core_module.c nginx-1.31.4/src/http/ngx_http_core_module.c
--- nginx-1.31.4/src/http/ngx_http_core_module.c	2026-08-20 18:25:06.301997145 +0000
+++ nginx-1.31.4/src/http/ngx_http_core_module.c	2026-08-20 18:27:15.262749710 +0000
@@ -55,7 +55,6 @@
     void *conf);
 static char *ngx_http_core_server_name(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
-static char *ngx_http_core_root(ngx_conf_t *cf, ngx_command_t *cmd, void *conf);
 static char *ngx_http_core_limit_except(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
 static char *ngx_http_core_set_aio(ngx_conf_t *cf, ngx_command_t *cmd,
@@ -64,6 +63,8 @@
     void *conf);
 static char *ngx_http_core_error_page(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
+static char *ngx_http_core_no_error_pages(ngx_conf_t *cf, ngx_command_t *cmd,
+    void *conf);
 static char *ngx_http_core_open_file_cache(ngx_conf_t *cf, ngx_command_t *cmd,
     void *conf);
 static char *ngx_http_core_error_log(ngx_conf_t *cf, ngx_command_t *cmd,
@@ -330,21 +331,6 @@
       offsetof(ngx_http_core_loc_conf_t, default_type),
       NULL },

-    { ngx_string("root"),
-      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_HTTP_LIF_CONF
-                        |NGX_CONF_TAKE1,
-      ngx_http_core_root,
-      NGX_HTTP_LOC_CONF_OFFSET,
-      0,
-      NULL },
-
-    { ngx_string("alias"),
-      NGX_HTTP_LOC_CONF|NGX_CONF_TAKE1,
-      ngx_http_core_root,
-      NGX_HTTP_LOC_CONF_OFFSET,
-      0,
-      NULL },
-
     { ngx_string("limit_except"),
       NGX_HTTP_LOC_CONF|NGX_CONF_BLOCK|NGX_CONF_1MORE,
       ngx_http_core_limit_except,
@@ -692,6 +678,14 @@
       0,
       NULL },

+    { ngx_string("no_error_pages"),
+      NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_HTTP_LIF_CONF
+                        |NGX_CONF_NOARGS,
+      ngx_http_core_no_error_pages,
+      NGX_HTTP_LOC_CONF_OFFSET,
+      0,
+      NULL },
+
     { ngx_string("post_action"),
       NGX_HTTP_MAIN_CONF|NGX_HTTP_SRV_CONF|NGX_HTTP_LOC_CONF|NGX_HTTP_LIF_CONF
                         |NGX_CONF_TAKE1,
@@ -3653,7 +3647,6 @@
      *     clcf->types = NULL;
      *     clcf->default_type = { 0, NULL };
      *     clcf->error_log = NULL;
-     *     clcf->error_pages = NULL;
      *     clcf->client_body_path = NULL;
      *     clcf->regex = NULL;
      *     clcf->exact_match = 0;
@@ -3663,6 +3656,7 @@
      *     clcf->keepalive_disable = 0;
      */

+    clcf->error_pages = NGX_CONF_UNSET_PTR;
     clcf->client_max_body_size = NGX_CONF_UNSET;
     clcf->client_body_buffer_size = NGX_CONF_UNSET_SIZE;
     clcf->client_body_timeout = NGX_CONF_UNSET_MSEC;
@@ -3867,9 +3861,7 @@
         }
     }

-    if (conf->error_pages == NULL && prev->error_pages) {
-        conf->error_pages = prev->error_pages;
-    }
+    ngx_conf_merge_ptr_value(conf->error_pages, prev->error_pages, NULL);

     ngx_conf_merge_str_value(conf->default_type,
                               prev->default_type, "text/plain");
@@ -4596,108 +4588,6 @@
 }


-static char *
-ngx_http_core_root(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
-{
-    ngx_http_core_loc_conf_t *clcf = conf;
-
-    ngx_str_t                  *value;
-    ngx_int_t                   alias;
-    ngx_uint_t                  n;
-    ngx_http_script_compile_t   sc;
-
-    alias = (cmd->name.len == sizeof("alias") - 1) ? 1 : 0;
-
-    if (clcf->root.data) {
-
-        if ((clcf->alias != 0) == alias) {
-            return "is duplicate";
-        }
-
-        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
-                           "\"%V\" directive is duplicate, "
-                           "\"%s\" directive was specified earlier",
-                           &cmd->name, clcf->alias ? "alias" : "root");
-
-        return NGX_CONF_ERROR;
-    }
-
-    if (clcf->named && alias) {
-        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
-                           "the \"alias\" directive cannot be used "
-                           "inside the named location");
-
-        return NGX_CONF_ERROR;
-    }
-
-    value = cf->args->elts;
-
-    if (ngx_strstr(value[1].data, "$document_root")
-        || ngx_strstr(value[1].data, "${document_root}"))
-    {
-        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
-                           "the $document_root variable cannot be used "
-                           "in the \"%V\" directive",
-                           &cmd->name);
-
-        return NGX_CONF_ERROR;
-    }
-
-    if (ngx_strstr(value[1].data, "$realpath_root")
-        || ngx_strstr(value[1].data, "${realpath_root}"))
-    {
-        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
-                           "the $realpath_root variable cannot be used "
-                           "in the \"%V\" directive",
-                           &cmd->name);
-
-        return NGX_CONF_ERROR;
-    }
-
-    clcf->alias = alias ? clcf->name.len : 0;
-    clcf->root = value[1];
-
-    if (!alias && clcf->root.len > 0
-        && clcf->root.data[clcf->root.len - 1] == '/')
-    {
-        clcf->root.len--;
-    }
-
-    if (clcf->root.data[0] != '$') {
-        if (ngx_conf_full_name(cf->cycle, &clcf->root, 0) != NGX_OK) {
-            return NGX_CONF_ERROR;
-        }
-    }
-
-    n = ngx_http_script_variables_count(&clcf->root);
-
-    ngx_memzero(&sc, sizeof(ngx_http_script_compile_t));
-    sc.variables = n;
-
-#if (NGX_PCRE)
-    if (alias && clcf->regex) {
-        clcf->alias = NGX_MAX_SIZE_T_VALUE;
-        n = 1;
-    }
-#endif
-
-    if (n) {
-        sc.cf = cf;
-        sc.source = &clcf->root;
-        sc.lengths = &clcf->root_lengths;
-        sc.values = &clcf->root_values;
-        sc.complete_lengths = 1;
-        sc.complete_values = 1;
-
-        if (ngx_http_script_compile(&sc) != NGX_OK) {
-            return NGX_CONF_ERROR;
-        }
-    }
-
-    return NGX_CONF_OK;
-}
-
-
 static ngx_http_method_name_t  ngx_methods_names[] = {
     { (u_char *) "GET",       (uint32_t) ~NGX_HTTP_GET },
     { (u_char *) "HEAD",      (uint32_t) ~NGX_HTTP_HEAD },
@@ -4955,6 +4845,10 @@
     ngx_http_compile_complex_value_t   ccv;

     if (clcf->error_pages == NULL) {
+        return "conflicts with \"no_error_pages\"";
+    }
+
+    if (clcf->error_pages == NGX_CONF_UNSET_PTR) {
         clcf->error_pages = ngx_array_create(cf->pool, 4,
                                              sizeof(ngx_http_err_page_t));
         if (clcf->error_pages == NULL) {
@@ -5058,6 +4952,25 @@

     return NGX_CONF_OK;
 }
+
+
+static char *
+ngx_http_core_no_error_pages(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
+{
+    ngx_http_core_loc_conf_t *clcf = conf;
+
+    if (clcf->error_pages == NULL) {
+        return "is duplicate";
+    }
+
+    if (clcf->error_pages != NGX_CONF_UNSET_PTR) {
+        return "conflicts with \"error_page\"";
+    }
+
+    clcf->error_pages = NULL;
+
+    return NGX_CONF_OK;
+}


 static char *
diff -urN nginx-1.31.4/src/http/ngx_http_upstream.c nginx-1.31.4/src/http/ngx_http_upstream.c
--- nginx-1.31.4/src/http/ngx_http_upstream.c	2026-08-20 18:25:06.305153944 +0000
+++ nginx-1.31.4/src/http/ngx_http_upstream.c	2026-08-20 18:27:15.237139382 +0000
@@ -1606,6 +1606,11 @@
         return;
     }

+    if (rc >= NGX_HTTP_SPECIAL_RESPONSE) {
+        ngx_http_upstream_finalize_request(r, u, rc);
+        return;
+    }
+
     u->state->peer = u->peer.name;

 #if (NGX_HTTP_UPSTREAM_ZONE)
@@ -2421,7 +2426,7 @@

 #endif

-    if (u->header_sent && !u->conf->preserve_output) {
+    if (u->request_body_sent && !u->conf->preserve_output) {
         u->write_event_handler = ngx_http_upstream_dummy_handler;

         (void) ngx_handle_write_event(c->write, 0);
diff -urN nginx-1.31.4/src/http/ngx_http_upstream.h nginx-1.31.4/src/http/ngx_http_upstream.h
--- nginx-1.31.4/src/http/ngx_http_upstream.h	2026-08-20 18:25:06.297042038 +0000
+++ nginx-1.31.4/src/http/ngx_http_upstream.h	2026-08-20 18:26:02.668719378 +0000
@@ -467,4 +467,9 @@
 extern ngx_conf_bitmask_t  ngx_http_upstream_ignore_headers_masks[];


+#ifndef HAVE_BALANCER_STATUS_CODE_PATCH
+#define HAVE_BALANCER_STATUS_CODE_PATCH
+#endif
+
+
 #endif /* _NGX_HTTP_UPSTREAM_H_INCLUDED_ */
diff -urN nginx-1.31.4/src/os/unix/ngx_daemon.c nginx-1.31.4/src/os/unix/ngx_daemon.c
--- nginx-1.31.4/src/os/unix/ngx_daemon.c	2026-08-20 18:25:06.335096609 +0000
+++ nginx-1.31.4/src/os/unix/ngx_daemon.c	2026-08-20 18:26:02.616133331 +0000
@@ -23,6 +23,8 @@
         break;

     default:
+        /* just to make it ASAN or Valgrind clean */
+        ngx_destroy_pool(ngx_cycle->pool);
         exit(0);
     }

diff -urN nginx-1.31.4/src/os/unix/ngx_process.c nginx-1.31.4/src/os/unix/ngx_process.c
--- nginx-1.31.4/src/os/unix/ngx_process.c	2026-08-20 18:25:06.334153155 +0000
+++ nginx-1.31.4/src/os/unix/ngx_process.c	2026-08-20 18:27:15.179015930 +0000
@@ -362,8 +362,15 @@
             break;

         case ngx_signal_value(NGX_RECONFIGURE_SIGNAL):
-            ngx_reconfigure = 1;
-            action = ", reconfiguring";
+            if (ngx_process == NGX_PROCESS_SINGLE) {
+                ngx_terminate = 1;
+                action = ", exiting";
+
+            } else {
+                ngx_reconfigure = 1;
+                action = ", reconfiguring";
+            }
+
             break;

         case ngx_signal_value(NGX_REOPEN_SIGNAL):
diff -urN nginx-1.31.4/src/os/unix/ngx_process_cycle.c nginx-1.31.4/src/os/unix/ngx_process_cycle.c
--- nginx-1.31.4/src/os/unix/ngx_process_cycle.c	2026-08-20 18:25:06.334321827 +0000
+++ nginx-1.31.4/src/os/unix/ngx_process_cycle.c	2026-08-20 18:27:15.361104636 +0000
@@ -15,6 +15,8 @@
     ngx_int_t type);
 static void ngx_start_cache_manager_processes(ngx_cycle_t *cycle,
     ngx_uint_t respawn);
+static void ngx_start_privileged_agent_processes(ngx_cycle_t *cycle,
+    ngx_uint_t respawn);
 static void ngx_pass_open_channel(ngx_cycle_t *cycle);
 static void ngx_signal_worker_processes(ngx_cycle_t *cycle, int signo);
 static ngx_uint_t ngx_reap_children(ngx_cycle_t *cycle);
@@ -24,6 +26,7 @@
 static void ngx_worker_process_exit(ngx_cycle_t *cycle);
 static void ngx_channel_handler(ngx_event_t *ev);
 static void ngx_cache_manager_process_cycle(ngx_cycle_t *cycle, void *data);
+static void ngx_privileged_agent_process_cycle(ngx_cycle_t *cycle, void *data);
 static void ngx_cache_manager_process_handler(ngx_event_t *ev);
 static void ngx_cache_loader_process_handler(ngx_event_t *ev);

@@ -52,6 +55,8 @@
 ngx_uint_t    ngx_noaccepting;
 ngx_uint_t    ngx_restart;

+ngx_uint_t    ngx_is_privileged_agent;
+

 static u_char  master_process[] = "master process";

@@ -130,6 +135,7 @@
     ngx_start_worker_processes(cycle, ccf->worker_processes,
                                NGX_PROCESS_RESPAWN);
     ngx_start_cache_manager_processes(cycle, 0);
+    ngx_start_privileged_agent_processes(cycle, 0);

     ngx_new_binary = 0;
     delay = 0;
@@ -215,6 +221,7 @@
                 ngx_start_worker_processes(cycle, ccf->worker_processes,
                                            NGX_PROCESS_RESPAWN);
                 ngx_start_cache_manager_processes(cycle, 0);
+                ngx_start_privileged_agent_processes(cycle, 0);
                 ngx_noaccepting = 0;

                 continue;
@@ -234,6 +241,7 @@
             ngx_start_worker_processes(cycle, ccf->worker_processes,
                                        NGX_PROCESS_JUST_RESPAWN);
             ngx_start_cache_manager_processes(cycle, 1);
+            ngx_start_privileged_agent_processes(cycle, 1);

             /* allow new processes to start */
             ngx_msleep(100);
@@ -248,6 +256,7 @@
             ngx_start_worker_processes(cycle, ccf->worker_processes,
                                        NGX_PROCESS_RESPAWN);
             ngx_start_cache_manager_processes(cycle, 0);
+            ngx_start_privileged_agent_processes(cycle, 0);
             live = 1;
         }

@@ -295,11 +304,26 @@
     }

     for ( ;; ) {
+        if (ngx_exiting) {
+            if (ngx_event_no_timers_left() == NGX_OK) {
+                ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "exiting");
+
+                for (i = 0; cycle->modules[i]; i++) {
+                    if (cycle->modules[i]->exit_process) {
+                        cycle->modules[i]->exit_process(cycle);
+                    }
+                }
+
+                ngx_master_process_exit(cycle);
+            }
+        }
+
         ngx_log_debug0(NGX_LOG_DEBUG_EVENT, cycle->log, 0, "worker cycle");

         ngx_process_events_and_timers(cycle);

-        if (ngx_terminate || ngx_quit) {
+        if (ngx_terminate) {
+            ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "exiting");

             for (i = 0; cycle->modules[i]; i++) {
                 if (cycle->modules[i]->exit_process) {
@@ -310,6 +334,20 @@
             ngx_master_process_exit(cycle);
         }

+        if (ngx_quit) {
+            ngx_quit = 0;
+            ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0,
+                          "gracefully shutting down");
+            ngx_setproctitle("process is shutting down");
+
+            if (!ngx_exiting) {
+                ngx_exiting = 1;
+                ngx_set_shutdown_timer(cycle);
+                ngx_close_listening_sockets(cycle);
+                ngx_close_idle_connections(cycle);
+            }
+        }
+
         if (ngx_reconfigure) {
             ngx_reconfigure = 0;
             ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "reconfiguring");
@@ -393,6 +431,35 @@


 static void
+ngx_start_privileged_agent_processes(ngx_cycle_t *cycle, ngx_uint_t respawn)
+{
+    ngx_core_conf_t       *ccf;
+
+    ccf = (ngx_core_conf_t *) ngx_get_conf(cycle->conf_ctx,
+                                           ngx_core_module);
+
+    if (!ccf->privileged_agent) {
+        return;
+    }
+
+    /* 0 is an illegal value and may result in a core dump later */
+    if (ccf->privileged_agent_connections == 0) {
+        ngx_log_error(NGX_LOG_ALERT, cycle->log, 0,
+                      "%ui worker_connection is not enough, "
+                      "privileged agent process cannot be spawned",
+                      ccf->privileged_agent_connections);
+        return;
+    }
+
+    ngx_spawn_process(cycle, ngx_privileged_agent_process_cycle,
+                      "privileged agent process", "privileged agent process",
+                      respawn ? NGX_PROCESS_JUST_RESPAWN : NGX_PROCESS_RESPAWN);
+
+    ngx_pass_open_channel(cycle);
+}
+
+
+static void
 ngx_pass_open_channel(ngx_cycle_t *cycle)
 {
     ngx_int_t      i;
@@ -689,6 +756,11 @@
     ngx_exit_cycle.files_n = ngx_cycle->files_n;
     ngx_cycle = &ngx_exit_cycle;

+    if (saved_init_cycle_pool != NULL && saved_init_cycle_pool != cycle->pool) {
+        ngx_destroy_pool(saved_init_cycle_pool);
+        saved_init_cycle_pool = NULL;
+    }
+
     ngx_destroy_pool(cycle->pool);

     exit(0);
@@ -796,7 +868,10 @@
         }
     }

-    if (geteuid() == 0) {
+    /*
+     * privileged agent process has the same permission as master process
+     */
+    if (!ngx_is_privileged_agent && geteuid() == 0) {
         if (setgid(ccf->group) == -1) {
             ngx_log_error(NGX_LOG_EMERG, cycle->log, ngx_errno,
                           "setgid(%d) failed", ccf->group);
@@ -1122,7 +1197,49 @@

         if (ngx_terminate || ngx_quit) {
             ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "exiting");
-            exit(0);
+            ngx_worker_process_exit(cycle);
+        }
+
+        if (ngx_reopen) {
+            ngx_reopen = 0;
+            ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "reopening logs");
+            ngx_reopen_files(cycle, -1);
+        }
+
+        ngx_process_events_and_timers(cycle);
+    }
+}
+
+
+static void
+ngx_privileged_agent_process_cycle(ngx_cycle_t *cycle, void *data)
+{
+    char   *name = data;
+    ngx_core_conf_t *ccf = (ngx_core_conf_t *) ngx_get_conf(cycle->conf_ctx, ngx_core_module);
+
+    /*
+     * Set correct process type since closing listening Unix domain socket
+     * in a master process also removes the Unix domain socket file.
+     */
+    ngx_process = NGX_PROCESS_HELPER;
+    ngx_is_privileged_agent = 1;
+
+    ngx_close_listening_sockets(cycle);
+
+    /* Set a moderate number of connections for a helper process. */
+    cycle->connection_n = ccf->privileged_agent_connections;
+
+    ngx_worker_process_init(cycle, -1);
+
+    ngx_use_accept_mutex = 0;
+
+    ngx_setproctitle(name);
+
+    for ( ;; ) {
+
+        if (ngx_terminate || ngx_quit) {
+            ngx_log_error(NGX_LOG_NOTICE, cycle->log, 0, "exiting");
+            ngx_worker_process_exit(cycle);
         }

         if (ngx_reopen) {
diff -urN nginx-1.31.4/src/os/unix/ngx_process_cycle.h nginx-1.31.4/src/os/unix/ngx_process_cycle.h
--- nginx-1.31.4/src/os/unix/ngx_process_cycle.h	2026-08-20 18:25:06.328260863 +0000
+++ nginx-1.31.4/src/os/unix/ngx_process_cycle.h	2026-08-20 18:26:02.716661213 +0000
@@ -45,6 +45,7 @@
 extern ngx_uint_t      ngx_inherited;
 extern ngx_uint_t      ngx_daemonized;
 extern ngx_uint_t      ngx_exiting;
+extern ngx_uint_t      ngx_is_privileged_agent;

 extern sig_atomic_t    ngx_reap;
 extern sig_atomic_t    ngx_sigio;
diff -urN nginx-1.31.4/src/os/win32/ngx_event_log.c nginx-1.31.4/src/os/win32/ngx_event_log.c
--- nginx-1.31.4/src/os/win32/ngx_event_log.c	2026-08-20 18:25:06.337301223 +0000
+++ nginx-1.31.4/src/os/win32/ngx_event_log.c	2026-08-20 18:25:11.557695096 +0000
@@ -8,7 +8,9 @@
 #include <ngx_core.h>


-#define NGX_MAX_ERROR_STR   2048
+#ifndef NGX_MAX_ERROR_STR
+#define NGX_MAX_ERROR_STR   4096
+#endif


 void ngx_cdecl
diff -urN nginx-1.31.4/src/stream/ngx_stream.h nginx-1.31.4/src/stream/ngx_stream.h
--- nginx-1.31.4/src/stream/ngx_stream.h	2026-08-20 18:25:06.288942082 +0000
+++ nginx-1.31.4/src/stream/ngx_stream.h	2026-08-20 18:26:02.669206636 +0000
@@ -27,6 +27,7 @@


 #define NGX_STREAM_OK                        200
+#define NGX_STREAM_SPECIAL_RESPONSE          300
 #define NGX_STREAM_BAD_REQUEST               400
 #define NGX_STREAM_FORBIDDEN                 403
 #define NGX_STREAM_INTERNAL_SERVER_ERROR     500
@@ -312,6 +313,15 @@
 } ngx_stream_module_t;


+typedef struct {
+    ngx_msec_t                       connect_timeout;
+    ngx_msec_t                       timeout;
+} ngx_stream_proxy_ctx_t;
+
+
+#define NGX_STREAM_HAVE_PROXY_TIMEOUT_FIELDS_PATCH 1
+
+
 #define NGX_STREAM_MODULE       0x4d525453     /* "STRM" */

 #define NGX_STREAM_MAIN_CONF    0x02000000
@@ -372,6 +382,7 @@
 extern ngx_module_t  ngx_stream_module;
 extern ngx_uint_t    ngx_stream_max_module;
 extern ngx_module_t  ngx_stream_core_module;
+extern ngx_module_t  ngx_stream_proxy_module;


 typedef ngx_int_t (*ngx_stream_filter_pt)(ngx_stream_session_t *s,
@@ -381,4 +392,7 @@
 extern ngx_stream_filter_pt  ngx_stream_top_filter;


+#define HAS_NGX_STREAM_PROXY_GET_NEXT_UPSTREAM_TRIES_PATCH 1
+
+
 #endif /* _NGX_STREAM_H_INCLUDED_ */
diff -urN nginx-1.31.4/src/stream/ngx_stream_proxy_module.c nginx-1.31.4/src/stream/ngx_stream_proxy_module.c
--- nginx-1.31.4/src/stream/ngx_stream_proxy_module.c	2026-08-20 18:25:06.293661195 +0000
+++ nginx-1.31.4/src/stream/ngx_stream_proxy_module.c	2026-08-20 18:26:02.670829171 +0000
@@ -449,6 +449,7 @@
     ngx_stream_proxy_srv_conf_t      *pscf;
     ngx_stream_upstream_srv_conf_t   *uscf, **uscfp;
     ngx_stream_upstream_main_conf_t  *umcf;
+    ngx_stream_proxy_ctx_t           *pctx;

     c = s->connection;

@@ -457,6 +458,17 @@
     ngx_log_debug0(NGX_LOG_DEBUG_STREAM, c->log, 0,
                    "proxy connection handler");

+    pctx = ngx_palloc(c->pool, sizeof(ngx_stream_proxy_ctx_t));
+    if (pctx == NULL) {
+        ngx_stream_proxy_finalize(s, NGX_STREAM_INTERNAL_SERVER_ERROR);
+        return;
+    }
+
+    pctx->connect_timeout = pscf->connect_timeout;
+    pctx->timeout = pscf->timeout;
+
+    ngx_stream_set_ctx(s, pctx, ngx_stream_proxy_module);
+
     u = ngx_pcalloc(c->pool, sizeof(ngx_stream_upstream_t));
     if (u == NULL) {
         ngx_stream_proxy_finalize(s, NGX_STREAM_INTERNAL_SERVER_ERROR);
@@ -758,6 +770,7 @@
     ngx_connection_t             *c, *pc;
     ngx_stream_upstream_t        *u;
     ngx_stream_proxy_srv_conf_t  *pscf;
+    ngx_stream_proxy_ctx_t       *ctx;

     c = s->connection;

@@ -765,6 +778,8 @@

     pscf = ngx_stream_get_module_srv_conf(s, ngx_stream_proxy_module);

+    ctx = ngx_stream_get_module_ctx(s, ngx_stream_proxy_module);
+
     u = s->upstream;

     u->connected = 0;
@@ -797,6 +812,11 @@
         return;
     }

+    if (rc >= NGX_STREAM_SPECIAL_RESPONSE) {
+        ngx_stream_proxy_finalize(s, rc);
+        return;
+    }
+
     u->state->peer = u->peer.name;

 #if (NGX_STREAM_UPSTREAM_ZONE)
@@ -847,7 +867,7 @@
     pc->read->handler = ngx_stream_proxy_connect_handler;
     pc->write->handler = ngx_stream_proxy_connect_handler;

-    ngx_add_timer(pc->write, pscf->connect_timeout);
+    ngx_add_timer(pc->write, ctx->connect_timeout);
 }


@@ -1029,13 +1049,15 @@
 static ngx_int_t
 ngx_stream_proxy_send_proxy_protocol(ngx_stream_session_t *s)
 {
-    u_char                       *p;
-    size_t                        size;
-    ssize_t                       n;
-    ngx_connection_t             *c, *pc;
-    ngx_stream_upstream_t        *u;
-    ngx_stream_proxy_srv_conf_t  *pscf;
-    static u_char                 buf[NGX_PROXY_PROTOCOL_MAX_HEADER];
+    u_char                  *p;
+    size_t                   size;
+    ssize_t                  n;
+    ngx_connection_t        *c, *pc;
+    ngx_stream_upstream_t   *u;
+    ngx_stream_proxy_ctx_t  *ctx;
+    static u_char            buf[NGX_PROXY_PROTOCOL_MAX_HEADER];
+
+    ctx = ngx_stream_get_module_ctx(s, ngx_stream_proxy_module);

     c = s->connection;

@@ -1068,9 +1090,7 @@
             return NGX_ERROR;
         }

-        pscf = ngx_stream_get_module_srv_conf(s, ngx_stream_proxy_module);
-
-        ngx_add_timer(pc->write, pscf->timeout);
+        ngx_add_timer(pc->write, ctx->timeout);

         pc->write->handler = ngx_stream_proxy_connect_handler;

@@ -1294,6 +1314,9 @@
     ngx_connection_t             *pc;
     ngx_stream_upstream_t        *u;
     ngx_stream_proxy_srv_conf_t  *pscf;
+    ngx_stream_proxy_ctx_t       *ctx;
+
+    ctx = ngx_stream_get_module_ctx(s, ngx_stream_proxy_module);

     u = s->upstream;

@@ -1349,7 +1372,7 @@
     if (rc == NGX_AGAIN) {

         if (!pc->write->timer_set) {
-            ngx_add_timer(pc->write, pscf->connect_timeout);
+            ngx_add_timer(pc->write, ctx->connect_timeout);
         }

         pc->ssl->handler = ngx_stream_proxy_ssl_handshake;
@@ -1744,6 +1767,7 @@
     ngx_stream_session_t         *s;
     ngx_stream_upstream_t        *u;
     ngx_stream_proxy_srv_conf_t  *pscf;
+    ngx_stream_proxy_ctx_t       *ctx;

     c = ev->data;
     s = c->data;
@@ -1755,6 +1779,8 @@
         return;
     }

+    ctx = ngx_stream_get_module_ctx(s, ngx_stream_proxy_module);
+
     c = s->connection;
     pc = u->peer.connection;

@@ -1774,7 +1800,7 @@
                 }

                 if (u->connected && !c->read->delayed && !pc->read->delayed) {
-                    ngx_add_timer(c->write, pscf->timeout);
+                    ngx_add_timer(c->write, ctx->timeout);
                 }

                 return;
@@ -1936,6 +1962,9 @@
     ngx_log_handler_pt            handler;
     ngx_stream_upstream_t        *u;
     ngx_stream_proxy_srv_conf_t  *pscf;
+    ngx_stream_proxy_ctx_t       *ctx;
+
+    ctx = ngx_stream_get_module_ctx(s, ngx_stream_proxy_module);

     u = s->upstream;

@@ -2132,7 +2161,7 @@
         }

         if (!c->read->delayed && !pc->read->delayed) {
-            ngx_add_timer(c->write, pscf->timeout);
+            ngx_add_timer(c->write, ctx->timeout);

         } else if (c->write->timer_set) {
             ngx_del_timer(c->write);
@@ -2902,3 +2931,14 @@

     return NGX_CONF_OK;
 }
+
+
+ngx_uint_t
+ngx_stream_proxy_get_next_upstream_tries(ngx_stream_session_t *s)
+{
+    ngx_stream_proxy_srv_conf_t      *pscf;
+
+    pscf = ngx_stream_get_module_srv_conf(s, ngx_stream_proxy_module);
+
+    return pscf->next_upstream_tries;
+}
diff -urN nginx-1.31.4/src/stream/ngx_stream_ssl_preread_module.c nginx-1.31.4/src/stream/ngx_stream_ssl_preread_module.c
--- nginx-1.31.4/src/stream/ngx_stream_ssl_preread_module.c	2026-08-20 18:25:06.291364190 +0000
+++ nginx-1.31.4/src/stream/ngx_stream_ssl_preread_module.c	2026-08-20 18:26:02.572882520 +0000
@@ -190,11 +190,11 @@
         }

         if (rc == NGX_OK) {
-            return ngx_stream_ssl_preread_servername(s, &ctx->host);
+            rc = ngx_stream_ssl_preread_servername(s, &ctx->host);
         }

         if (rc != NGX_AGAIN) {
-            return rc;
+            return rc == NGX_OK ? NGX_DECLINED : rc;
         }

         p += len;
diff -urN nginx-1.31.4/src/stream/ngx_stream_upstream.h nginx-1.31.4/src/stream/ngx_stream_upstream.h
--- nginx-1.31.4/src/stream/ngx_stream_upstream.h	2026-08-20 18:25:06.290171871 +0000
+++ nginx-1.31.4/src/stream/ngx_stream_upstream.h	2026-08-20 18:26:02.671362699 +0000
@@ -166,4 +166,9 @@
 extern ngx_module_t  ngx_stream_upstream_module;


+#ifndef HAVE_BALANCER_STATUS_CODE_PATCH
+#define HAVE_BALANCER_STATUS_CODE_PATCH
+#endif
+
+
 #endif /* _NGX_STREAM_UPSTREAM_H_INCLUDED_ */
diff -urN nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.c nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.c
--- nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.c	2026-08-20 18:25:06.292993715 +0000
+++ nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.c	2026-08-20 18:25:11.572327344 +0000
@@ -21,10 +21,6 @@

 #if (NGX_STREAM_SSL)

-static ngx_int_t ngx_stream_upstream_set_round_robin_peer_session(
-    ngx_peer_connection_t *pc, void *data);
-static void ngx_stream_upstream_save_round_robin_peer_session(
-    ngx_peer_connection_t *pc, void *data);
 static ngx_int_t ngx_stream_upstream_empty_set_session(
     ngx_peer_connection_t *pc, void *data);
 static void ngx_stream_upstream_empty_save_session(ngx_peer_connection_t *pc,
@@ -933,7 +929,7 @@

 #if (NGX_STREAM_SSL)

-static ngx_int_t
+ngx_int_t
 ngx_stream_upstream_set_round_robin_peer_session(ngx_peer_connection_t *pc,
     void *data)
 {
@@ -995,7 +991,7 @@
 }


-static void
+void
 ngx_stream_upstream_save_round_robin_peer_session(ngx_peer_connection_t *pc,
     void *data)
 {
diff -urN nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.h nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.h
--- nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.h	2026-08-20 18:25:06.288163640 +0000
+++ nginx-1.31.4/src/stream/ngx_stream_upstream_round_robin.h	2026-08-20 18:25:11.572718310 +0000
@@ -146,6 +146,16 @@
         ngx_rwlock_unlock(&peer->lock);                                       \
     }

+#if (NGX_STREAM_SSL)
+ngx_int_t ngx_stream_upstream_set_round_robin_peer_session(
+    ngx_peer_connection_t *pc, void *data);
+void ngx_stream_upstream_save_round_robin_peer_session(
+    ngx_peer_connection_t *pc, void *data);
+#endif
+
+
+#define HAVE_NGX_STREAM_BALANCER_EXPORT_PATCH 1
+

 #define ngx_stream_upstream_rr_peer_ref(peers, peer)                          \
     (peer)->refs++;
